File: Remove_duplicates_Linked_List_2/Remove_Duplicates_Linked_list_2.java
import java.util.HashMap;

class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    Node head;

    public void append(int data) {
        Node newNode = new Node(data);
        if(head == null) {
            head = newNode;
            return;
        }
        Node current = head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = newNode;
    }

    public void removeDuplicates() {
        if (head == null) return;

        HashMap<Integer, Integer> frequency = new HashMap<>();        Node current = head;
        while (current != null) {
            frequency.put(current.data, frequency.getOrDefault(current.data, 0) + 1);
            current = current.next;
        }

        Node dummy = new Node(0);
        dummy.next = head;
        Node prev = dummy;
        current = head;

        while (current != null) {
            if (frequency.get(current.data) > 1) {
                prev.next = current.next;
            }
            else {
                prev = current;
            }
            current = current.next;

            head = dummy.next;
        }
    }

    public void display() {
        if (head == null) {
            System.out.println("The list is empty");
            return;
        }

        Node current = head;
        while (current != null) {
            System.out.println(current.data + "->");
            current = current.next;
        }
        System.out.println("None");

    }

    public static void main(String[] args) {
        LinkedList ll = new LinkedList();
        ll.append(10);
        ll.append(20);
        ll.append(10);
        ll.append(30);
        ll.append(20);

        System.out.println("Original Linked List");
        ll.display();

        ll.removeDuplicates();

        System.out.println("List after Removing Duplicates:");
        ll.display();
    }
}

File: Remove_duplicates_Linked_List_2/Remove_Duplicates_Linked_list_2.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    def remove_duplicates(self):
        if not self.head:
            return

        # Step 1: Count frequencies of each value
        frequency = {}
        current = self.head
        while current:
            frequency[current.data] = frequency.get(current.data, 0) + 1
            current = current.next

        # Step 2: Rebuild the list, skipping values with frequency > 1
        dummy = Node(0)  # Temporary dummy node to simplify edge cases
        dummy.next = self.head
        prev = dummy
        current = self.head

        while current:
            if frequency[current.data] > 1:  # Skip nodes with duplicates
                prev.next = current.next
            else:
                prev = current
            current = current.next

        self.head = dummy.next  # Update head to skip any duplicates at the start

    def display(self):
        current = self.head
        while current:
            print(current.data, end="->")
            current = current.next
        print("None")

ll = LinkedList()
ll.append(10)
ll.append(20)
ll.append(10)
ll.append(30)
ll.append(20)

ll.display()
ll.remove_duplicates()
ll.display()

